core/src/commonMain/kotlin/com/boswelja/ephemeris/core/ui/CalendarPageLoader.kt

```diff
@@ -0,0 +1,123 @@
+package com.boswelja.ephemeris.core.ui
+
+import com.boswelja.ephemeris.core.data.CalendarPageSource
+import com.boswelja.ephemeris.core.data.FocusMode
+import com.boswelja.ephemeris.core.model.DisplayDate
+import com.boswelja.ephemeris.core.model.DisplayRow
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.launch
+import kotlinx.datetime.LocalDate
+import kotlin.math.abs
+
+/**
+ * Handles loading data from [CalendarPageSource], combining with [FocusMode], caching and
+ * prefetching entries. Platform UIs should make use of this class for their data loading. If the
+ * page source or focus mode change, it is expected a new instance will be created and the existing
+ * instance discarded. Prefetch and cache operations are handled asynchronously via [coroutineScope].
+ */
+@OptIn(FlowPreview::class)
+public class CalendarPageLoader(
+    private val coroutineScope: CoroutineScope,
+    public val calendarPageSource: CalendarPageSource,
+    public val focusMode: FocusMode
+) {
+    private val pageCache = mutableMapOf<Int, Set<DisplayRow>>()
+
+    private val lastLoadedPage = MutableStateFlow(0)
+
+    init {
+        // Launch the page cache job
+        coroutineScope.launch {
+            lastLoadedPage
+                .debounce(PAGE_CHANGE_DEBOUNCE_MILLIS) // Debounce here to reduce overlapping jobs
+                .collect {
+                    if (tryBuildCache(it)) {
```

Comment:
I'm tempted to suggest
```
lastLoadedPage
    .filter { tryBuildCache(it) }
    .debounce(PAGE_CHANGE_DEBOUNCE_MILLIS)
    .collect { trimCache(it) }
```
to reduce emissions to the collector? And obviously I'd prefer onEach + launchIn for code line alignment

Reply:
filter would be called before debounce though right? Effectively making debounce redundant

Reply:
Probably need something more robust than debounce in general on second though, since there's a chance it will block emissions for too long

Reply:
is the debounce guarding both the building and the trimming of the cache then? Do we need a mutex to lock/unlock operations to avoid overlapping?

Reply:
Lock could work, debounce is there to guard both yeah. It *shouldn't* be necessary when we consider the average user scroll speed

Reply:
Maybe I'll just remove debounce for now. If it causes issues down the line we can come back to it